docs/topics/db/queries.txt

```diff
@@ -1009,6 +1009,65 @@ Unless you are sure you wish to work with SQL ``NULL`` values, consider setting
 Key, index, and path transforms
 -------------------------------
 
+.. class:: json.KeyTransform(key_name, *args, **kwargs)
+
+A :class:`~django.db.models.Transform` to access a key named ``key_name`` in a
+:class:`~django.db.models.JSONField`. ``key_name`` could also be an integer to
+transform to an array index. The field instance which is to be transformed
+should be passed as the second argument. This would return the resulting key
+transform which is also in the form of json.
+
+To query on a :class:`~django.db.models.JSONField`::
+
+    >>> Dog.objects.create(name='Rufus', data={
+    ...     'breed': 'labrador',
+    ...     'owner': {
+    ...         'name': 'Bob',
+    ...         'other_pets': [{
+    ...             'name': 'Fishy',
+    ...         }],
+    ...     },
+    ... })
+    <Dog: Rufus>
+    >>> Dog.objects.annotate(
+    ...     breed_name=KeyTransform('breed', 'data')
+    ... ).filter(breed_name__exact='labrador')
+    <QuerySet [<Dog: Rufus>]>
+
+Nested key transforms can also be done as follows::
+
+    >>> Dogs.objects.annotate(
+    ...     owner_other_pet_name=KeyTransform(
+    ...         'other_pets',KeyTransform('owner', 'data')
+    ...     )).filter(owner_other_pet_name__exact=[{'name': 'Fishy'}])
+    <QuerySet [<Dog: Rufus>]>
+
+To query the resulting json as text, the :class:`~django.db.models.functions.Cast`
+function needs to be used. This is done as follows::
+
+    >>> Dogs.objects.annotate(
+    ...     breed_name=Cast(KeyTransform('breed', 'data'), TextField())
+    ... ).filter(breed_name__contains='lab')
+    <QuerySet [<Dog: Rufus>]>
+
+.. class:: json.KeyTextTransform(key_name, *args, **kwargs)
+
+    This returns the resulting key transform in the form of text. This would also
+    need to be :class:`~django.db.models.functions.Cast` to a ``TextField`` to be
+    treat the result as text.
+
+.. warning::
+
+    Casting a resulting KeyTransform might still be in resemble json format.
+    For example, ``Cast(KeyTransform('breed', 'data'), TextField())`` would result
+    in ``'"labrador"'`` instead of ``'labrador'``. So filtering
+    ``breed_name__contains='dor"'`` would also be true.
+
+.. note::
+
+    .. class:: json.KeyTextTransform is alias to .. class:: json.KeyTransform for
+    non-postgresql databases.
```

Comment:
Is this an oversight? I'm surprised to learn that.

Should we make sure that `KeyTextTransform` actually results in `text` (or whatever database dependent type is used to represent strings) before making it public?

It seems [the `->>` operator is supported on MySQL 5.7+](https://dev.mysql.com/doc/refman/5.7/en/json-search-functions.html#operator_json-inline-path) and on other backend we could make sure that the return value is automatically wrapped in a `CAST`?

Reply:
It works on SQLite and Oracle as a side-effect of the ticket-32213. I've prepared a fix for MySQL, see #15970.